File: sdks/python/ans/broker.py
"""Identity Broker client for ephemeral credential provisioning."""

from typing import Dict, List, Optional
import time
from .client import ANSClient


class Scope:
    """Defines what a credential can access."""

    def __init__(
        self,
        resource: str,
        permissions: List[str],
        constraints: Optional[Dict[str, str]] = None,
    ):
        """
        Initialize a scope.
        
        Args:
            resource: Target resource (e.g., "s3://bucket/object", "vault://secret/data")
            permissions: Allowed operations (e.g., ["read"], ["write"], ["read", "write"])
            constraints: Additional restrictions (e.g., {"ip": "192.168.1.0/24"})
        """
        self.resource = resource
        self.permissions = permissions
        self.constraints = constraints or {}

    def to_dict(self) -> Dict:
        """Convert to dict for JSON serialization."""
        return {
            "resource": self.resource,
            "permissions": self.permissions,
            "constraints": self.constraints,
        }


class Credential:
    """An ephemeral, scoped credential."""

    def __init__(self, data: Dict):
        self.credential_id = data["credential_id"]
        self.agent_id = data["agent_id"]
        self.provider = data["provider"]
        self.type = data["type"]
        self.secret = data["secret"]
        self.metadata = data.get("metadata", {})
        self.scope = data["scope"]
        self.issued_at = data["issued_at"]
        self.expires_at = data["expires_at"]
        self.revoked = data.get("revoked", False)
        self.request_id = data.get("request_id", "")
        self.pre_receipt_id = data.get("pre_receipt_id", "")

    def __repr__(self) -> str:
        return (f"Credential(credential_id={self.credential_id!r}, "
                f"agent_id={self.agent_id!r}, provider={self.provider!r}, "
                f"type={self.type!r}, secret='[REDACTED]', "
                f"scope={self.scope!r}, issued_at={self.issued_at!r}, "
                f"expires_at={self.expires_at!r}, revoked={self.revoked})")

    def is_expired(self) -> bool:
        """Check if the credential has expired."""
        return time.time() > self.expires_at

    def is_active(self) -> bool:
        """Check if the credential is active (not expired and not revoked)."""
        return not self.is_expired() and not self.revoked


class IdentityBroker:
    """
    Client for the ANS Identity Broker.
    
    Provisions ephemeral, scoped credentials for zero-trust agent access.
    
    Example:
        from ans.broker import IdentityBroker, Scope
        
        broker = IdentityBroker(ans_client)
        
        # Provision a 60-second credential for S3 read access
        cred = broker.provision(
            provider="aws-iam",
            agent_id="ans_3vQb7uL6x9",
            action_type="file.read",
            scope=Scope(
                resource="s3://my-bucket/data.json",
                permissions=["read"]
            ),
            ttl_seconds=60,
            pre_receipt_id="abc123"
        )
        
        # Use the credential
        import boto3
        s3 = boto3.client(
            's3',
            aws_access_key_id=cred.secret,
            aws_secret_access_key=cred.metadata["secret_key"],
            aws_session_token=cred.metadata["session_token"]
        )
        
        # Revoke when done (or wait 60s for auto-expiry)
        broker.revoke(cred.credential_id)
    """

    def __init__(self, client: ANSClient):
        self.client = client

    def provision(
        self,
        provider: str,
        agent_id: str,
        action_type: str,
        scope: Scope,
        ttl_seconds: int = 60,
        pre_receipt_id: str = "",
        parent_agent_id: str = "",
    ) -> Credential:
        """
        Provision an ephemeral credential.
        
        Args:
            provider: Provider name ("vault", "aws-iam", "gcp-iam", etc.)
            agent_id: Agent requesting the credential
            action_type: Action type (e.g., "file.read", "http.post")
            scope: What the credential can access
            ttl_seconds: TTL in seconds (max 60)
            pre_receipt_id: Linked pre-action receipt ID
            parent_agent_id: Parent agent ID if sub-agent
        
        Returns:
            Credential object
        """
        ttl = min(ttl_seconds, 60)
        resp = self.client.token_request(
            agent_id=agent_id,
            resource=scope.resource,
            action=scope.permissions[0] if scope.permissions else "read",
            ttl_seconds=ttl,
            single_use=True,
        )
        data = {
            "credential_id": resp.get("token_id", ""),
            "agent_id": agent_id,
            "provider": provider,
            "type": resp.get("token_type", ""),
            "secret": resp.get("access_key", "") or resp.get("bearer_token", ""),
            "metadata": {
                "secret_key": resp.get("secret_key", ""),
                "session_token": resp.get("session_token", ""),
            },
            "scope": scope.to_dict(),
            "issued_at": 0,
            "expires_at": resp.get("expires_ns", 0) / 1e9 if resp.get("expires_ns") else 0,
        }
        return Credential(data)

    def revoke(self, credential_id: str) -> None:
        """
        Revoke a credential immediately.
        
        Args:
            credential_id: Credential ID to revoke
        """
        self.client.token_revoke(credential_id)

    def get(self, credential_id: str) -> Optional[Credential]:
        """
        Get a credential from the cache.
        
        Args:
            credential_id: Credential ID
        
        Returns:
            Credential object or None if not found
        """
        tokens = self.client.token_list()
        for t in tokens:
            if t.get("token_id") == credential_id:
                data = {
                    "credential_id": t["token_id"],
                    "agent_id": t.get("agent_id", ""),
                    "provider": t.get("provider", ""),
                    "type": t.get("token_type", ""),
                    "secret": "",
                    "metadata": {},
                    "scope": {"resource": t.get("resource", ""), "permissions": [t.get("action", "read")]},
                    "issued_at": t.get("created_ns", 0) / 1e9,
                    "expires_at": t.get("expires_ns", 0) / 1e9,
                    "revoked": t.get("state") != "active",
                }
                return Credential(data)
        return None

    def list_active(self) -> List[Credential]:
        """
        List all active (non-expired, non-revoked) credentials.
        
        Returns:
            List of Credential objects
        """
        tokens = self.client.token_list()
        result = []
        for t in tokens:
            if t.get("state") != "active":
                continue
            data = {
                "credential_id": t["token_id"],
                "agent_id": t.get("agent_id", ""),
                "provider": t.get("provider", ""),
                "type": t.get("token_type", ""),
                "secret": "",
                "metadata": {},
                "scope": {"resource": t.get("resource", ""), "permissions": [t.get("action", "read")]},
                "issued_at": t.get("created_ns", 0) / 1e9,
                "expires_at": t.get("expires_ns", 0) / 1e9,
            }
            result.append(Credential(data))
        return result


# Context manager for automatic credential provisioning and revocation
class ephemeral_credential:
    """
    Context manager for automatic credential lifecycle management.
    
    Example:
        from ans.broker import IdentityBroker, Scope, ephemeral_credential
        
        broker = IdentityBroker(ans_client)
        
        with ephemeral_credential(
            broker,
            provider="aws-iam",
            agent_id="ans_xyz",
            action_type="file.read",
            scope=Scope("s3://bucket/file", ["read"])
        ) as cred:
            # Use credential for 60 seconds
            s3_client = boto3.client('s3', aws_access_key_id=cred.secret, ...)
            data = s3_client.get_object(Bucket='bucket', Key='file')
        # Credential is automatically revoked here
    """

    def __init__(
        self,
        broker: IdentityBroker,
        provider: str,
        agent_id: str,
        action_type: str,
        scope: Scope,
        ttl_seconds: int = 60,
        pre_receipt_id: str = "",
    ):
        self.broker = broker
        self.provider = provider
        self.agent_id = agent_id
        self.action_type = action_type
        self.scope = scope
        self.ttl_seconds = ttl_seconds
        self.pre_receipt_id = pre_receipt_id
        self.credential = None

    def __enter__(self) -> Credential:
        """Provision the credential."""
        self.credential = self.broker.provision(
            provider=self.provider,
            agent_id=self.agent_id,
            action_type=self.action_type,
            scope=self.scope,
            ttl_seconds=self.ttl_seconds,
            pre_receipt_id=self.pre_receipt_id,
        )
        return self.credential

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Revoke the credential."""
        if self.credential:
            try:
                self.broker.revoke(self.credential.credential_id)
            except Exception:
                pass  # Best-effort revocation

File: internal/broker/broker.go
// Package broker — Ephemeral Identity Provisioning for Zero-Trust Agent Access.
//
// Agents execute under the developer's admin keys by default — massive insider threat.
// The Identity Broker provisions ephemeral, scoped, single-use credentials with 60-second TTLs.
//
// Supported backends:
// - HashiCorp Vault (dynamic secrets)
// - AWS IAM (temporary credentials via STS AssumeRole)
// - Google Cloud IAM (short-lived service account tokens)
// - Azure AD (ephemeral app registrations)
// - Generic OAuth2/OIDC (client credentials flow with TTL)
//
// SPDX-License-Identifier: MIT
package broker

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

// Provider defines the interface for credential providers (Vault, AWS, GCP, etc.)
type Provider interface {
	// Name returns the provider name (e.g., "vault", "aws-iam", "gcp-iam")
	Name() string

	// ProvisionCredential generates an ephemeral credential for the given scope.
	// Returns a Credential with a 60-second TTL (or provider-defined max TTL).
	ProvisionCredential(ctx context.Context, req *ProvisionRequest) (*Credential, error)

	// RevokeCredential immediately revokes the credential (best-effort).
	RevokeCredential(ctx context.Context, credentialID string) error

	// ValidateScope checks if the scope is valid for this provider.
	ValidateScope(scope Scope) error
}

// ProvisionRequest describes what access the agent needs.
type ProvisionRequest struct {
	AgentID       string    `json:"agent_id"`        // Agent requesting the credential
	ActionType    string    `json:"action_type"`     // e.g., "file.read", "http.post", "db.query"
	Scope         Scope     `json:"scope"`           // What resources the agent can access
	TTLSeconds    int       `json:"ttl_seconds"`     // Requested TTL (max 60s enforced)
	RequestID     string    `json:"request_id"`      // Unique request ID for audit
	PreReceiptID  string    `json:"pre_receipt_id"`  // Linked pre-action receipt
	ParentAgentID string    `json:"parent_agent_id"` // Parent agent if sub-agent
	Timestamp     time.Time `json:"timestamp"`       // Request timestamp
}

// Scope defines what the credential can access.
type Scope struct {
	// Resource is the target (e.g., "s3://bucket/object", "db://prod/users", "file:///etc/nginx/conf")
	Resource string `json:"resource"`

	// Permissions is the set of allowed operations (e.g., ["read"], ["write"], ["read", "write"])
	Permissions []string `json:"permissions"`

	// Constraints are additional restrictions (e.g., "ip:192.168.1.0/24", "time:14:00-15:00")
	Constraints map[string]string `json:"constraints,omitempty"`
}

// Credential is an ephemeral, scoped credential.
type Credential struct {
	CredentialID   string            `json:"credential_id"`   // Unique credential ID
	AgentID        string            `json:"agent_id"`        // Agent that owns this credential
	ProviderName   string            `json:"provider"`        // Provider that issued it
	Type           string            `json:"type"`            // "aws-sts", "vault-token", "oauth2-bearer", etc.
	Secret         string            `json:"secret"`          // The actual secret (token, key, etc.)
	Metadata       map[string]string `json:"metadata"`        // Provider-specific metadata (e.g., "role_arn")
	Scope          Scope             `json:"scope"`           // What this credential can access
	IssuedAt       time.Time         `json:"issued_at"`       // When issued
	ExpiresAt      time.Time         `json:"expires_at"`      // When it expires
	Revoked        bool              `json:"revoked"`         // Whether it's been revoked
	RequestID      string            `json:"request_id"`      // Request ID that generated this
	PreReceiptID   string            `json:"pre_receipt_id"`  // Pre-action receipt ID
	ProvisionLogID string            `json:"provision_log_id"` // Log entry ID
}

// Broker manages credential provisioning and lifecycle.
type Broker struct {
	providers map[string]Provider // provider name → provider
	cache     *CredentialCache
	logger    Logger
	mu        sync.RWMutex
	done      chan struct{}
}

// NewBroker creates a new identity broker.
func NewBroker(logger Logger) *Broker {
	return &Broker{
		providers: make(map[string]Provider),
		cache:     NewCredentialCache(),
		logger:    logger,
		done:      make(chan struct{}),
	}
}

// RegisterProvider registers a credential provider.
func (b *Broker) RegisterProvider(p Provider) error {
	b.mu.Lock()
	defer b.mu.Unlock()

	name := p.Name()
	if _, exists := b.providers[name]; exists {
		return fmt.Errorf("provider %q already registered", name)
	}
	b.providers[name] = p
	b.logger.Infof("Registered credential provider: %s", name)
	return nil
}

// Provision generates an ephemeral credential.
func (b *Broker) Provision(ctx context.Context, providerName string, req *ProvisionRequest) (*Credential, error) {
	b.mu.RLock()
	provider, ok := b.providers[providerName]
	b.mu.RUnlock()

	if !ok {
		return nil, fmt.Errorf("provider %q not found", providerName)
	}

	// Enforce max TTL of 60 seconds
	if req.TTLSeconds > 60 || req.TTLSeconds <= 0 {
		req.TTLSeconds = 60
	}

	// Validate scope
	if err := provider.ValidateScope(req.Scope); err != nil {
		return nil, fmt.Errorf("invalid scope: %w", err)
	}

	// Generate request ID if not provided
	if req.RequestID == "" {
		req.RequestID = generateRequestID()
	}
	req.Timestamp = time.Now()

	// Provision the credential
	b.logger.Infof("Provisioning credential for agent %s (provider: %s, scope: %s)",
		req.AgentID, providerName, req.Scope.Resource)

	cred, err := provider.ProvisionCredential(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("provider %s failed to provision: %w", providerName, err)
	}

	// Cache the credential
	b.cache.Store(cred)

	// Schedule auto-revocation — no request context; background goroutine uses its own timeout
	go b.scheduleRevocation(cred) // #nosec G118

	b.logger.Infof("Credential %s issued for agent %s (expires in %ds)",
		cred.CredentialID, cred.AgentID, req.TTLSeconds)

	return cred, nil
}

// Revoke immediately revokes a credential.
func (b *Broker) Revoke(ctx context.Context, credentialID string) error {
	cred := b.cache.Get(credentialID)
	if cred == nil {
		return fmt.Errorf("credential %q not found", credentialID)
	}

	b.mu.RLock()
	provider, ok := b.providers[cred.ProviderName]
	b.mu.RUnlock()

	if !ok {
		return fmt.Errorf("provider %q not found", cred.ProviderName)
	}

	b.logger.Infof("Revoking credential %s (agent: %s)", credentialID, cred.AgentID)

	if err := provider.RevokeCredential(ctx, credentialID); err != nil {
		return fmt.Errorf("provider %s failed to revoke: %w", cred.ProviderName, err)
	}

	b.cache.MarkRevoked(credentialID) // Update cache with revoked status

	b.logger.Infof("Credential %s revoked", credentialID)
	return nil
}

// Get retrieves a credential from the cache.
func (b *Broker) Get(credentialID string) *Credential {
	return b.cache.Get(credentialID)
}

// ListActive returns all active (non-expired, non-revoked) credentials.
func (b *Broker) ListActive() []*Credential {
	return b.cache.ListActive()
}

// Close stops all pending revocation goroutines.
func (b *Broker) Close() {
	close(b.done)
}

// scheduleRevocation automatically revokes the credential when it expires.
func (b *Broker) scheduleRevocation(cred *Credential) {
	ttl := time.Until(cred.ExpiresAt)
	if ttl <= 0 {
		return
	}

	select {
	case <-b.done:
		return
	case <-time.After(ttl):
	}

	// Best-effort revocation; no request context available in background goroutine
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := b.Revoke(ctx, cred.CredentialID); err != nil {
		b.logger.Errorf("Auto-revoke failed for credential %s: %v", cred.CredentialID, err)
	}
}

// CredentialCache is a thread-safe cache for credentials.
type CredentialCache struct {
	creds map[string]*Credential
	mu    sync.RWMutex
}

func NewCredentialCache() *CredentialCache {
	return &CredentialCache{
		creds: make(map[string]*Credential),
	}
}

func (c *CredentialCache) Store(cred *Credential) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.creds[cred.CredentialID] = cred
}

func (c *CredentialCache) Get(credentialID string) *Credential {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.creds[credentialID]
}

func (c *CredentialCache) MarkRevoked(credentialID string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if cred, ok := c.creds[credentialID]; ok {
		cred.Revoked = true
	}
}

func (c *CredentialCache) ListActive() []*Credential {
	c.mu.RLock()
	defer c.mu.RUnlock()

	now := time.Now()
	active := make([]*Credential, 0)
	for _, cred := range c.creds {
		if !cred.Revoked && now.Before(cred.ExpiresAt) {
			active = append(active, cred)
		}
	}
	return active
}

// Logger interface for broker logging.
type Logger interface {
	Infof(format string, args ...interface{})
	Errorf(format string, args ...interface{})
}

// generateRequestID generates a unique request ID.
func generateRequestID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		// Fallback to nanosecond timestamp if crypto/rand fails
		return fmt.Sprintf("fallback-%d", time.Now().UnixNano())
	}
	h := sha256.Sum256(b)
	return hex.EncodeToString(h[:8])
}

// Sensitive metadata keys that should be redacted in JSON output.
var sensitiveMetaKeys = map[string]bool{
	"access_key":    true,
	"session_token": true,
	"secret_key":    true,
}

// MarshalJSON redacts the secret and sensitive metadata when marshaling to JSON.
func (c *Credential) MarshalJSON() ([]byte, error) {
	type Alias Credential
	redactedMeta := make(map[string]string, len(c.Metadata))
	for k, v := range c.Metadata {
		if sensitiveMetaKeys[k] {
			redactedMeta[k] = "[REDACTED]"
		} else {
			redactedMeta[k] = v
		}
	}
	return json.Marshal(&struct {
		Secret   string            `json:"secret"`
		Metadata map[string]string `json:"metadata"`
		*Alias
	}{
		Secret:   "[REDACTED]",
		Metadata: redactedMeta,
		Alias:    (*Alias)(c),
	})
}
